### PR TITLE
[Python] toggle_multiline comprehensions and "if/else <-> ternary"

### PR DESCRIPTION
**Pull request by alexpw**
Hi!

This plugin has been super helpful for me (thank you!), so I built out more python support.   I tried to keep commits isolated to `filetypes/python.lua` to avoid design decisions, but there are a few non-breaking changes outside of it.  I try to give context/explain the reason why for each (see below: Core Changes).

Features:
--
Comprehensions:
```python
# list_comprehension
xs = [x for x in range(10) if x - 3 or -x and x - 3 == 0 and abs(x - 1) < 2]
# set_comprehension
xs = {x for x in range(10)}
# dictionary_comprehension
xs = {x: 1 for x in range(10)}
# generator_expression
xs = (x for x in range(10))
# more complex example
xs = xs + bar([x for x in range(10) if x - 3 or -x and x - 3 == 0]) + xs
```
The result of toggle_multiline() on the complex example (cursor on the square bracket):
```python
xs = xs + bar([
                  x
                  for x in range(10)
                  if x - 3 or -x and x - 3 == 0
              ]) + xs
```

Inlined if statements:
```python
# assignment
x = 1 if foo() > 100 else 2
# multi assignment
x = y = z = 1 if foo() > 100 else 2
# return
return 1 if foo() > 100 else 2
# fn call
print(3) if True else print(4)
# lambda
x = (lambda y: y + 1) if foo() > 100 else (lambda y: y - 1)
# 1 line if_statement
if x: foo()
# 1 line if statement with ternary if
if x: foo() if y else bar()
# 1 line for statement with ternary if
for x in range(10): y = x + 1 if x < 2 else x - 1
```
are expanded (via _node_action on the `if`_) to:
```python
# assignment
if foo() > 100:
    x = 1
else:
    x = 2
# multi assignment
if foo() > 100:
    x = y = z = 1
else:
    x = y = z = 2
# return
if foo() > 100:
    return 1
else:
    return 2
# fn call
if True:
    print(3)
else:
    print(4)
# lambda
if foo() > 100:
    x = (lambda y: y + 1)
else:
    x = (lambda y: y - 1)
# 1 line if_statement
if x:
    foo()
# 1 line if statement with ternary if (cursor on first `if`)
if x:
    foo() if y else bar()
# 1 line if statement with ternary if (cursor on ternary `if`)
if x:
    if y:
        foo()
    else:
        bar()
# 1 line for statement with ternary if
for x in range(10):
    if x < 2:
        y = x + 1
    else:
        y = x - 1
```
These expanded forms can all be inlined, but when it can't safely do so or when it doesn't make sense to, it won't.

Core Changes
--
1. `init.lua:do_action()` - An optional `target_node` can be returned by actions and used as the target for replacement. 

This was used for python's "if/else <-> ternary", where the parent of the node being acted on is instead the one that needs to be replaced.  For example:
```python
x = 1 if y is not None and foo() > 100 else 2
```
The `if` is a "conditional_expression" that initiates the node_action.  Once expanded, the `if` is again the target for the inline_if_stmt node_action.
```python
if y is not None and foo() > 100:
    x = 1
else:
    x = 2
```
2. `init.lua:replace_node()` - The options returned by an action gained a `trim_whitespace` key that trims trailing whitespace between the `start_row` and `end_row`.  It is a table expecting keys by the same name, with relative values adjust the default ones, to match the cursor row/col design and behavior.

This was necessary to support expanding: 
```python
for x in range(10): y = 1 if x < 2 else x - 1
```
The expanded replacement needs to go on the next line and leaves behind 1 trailing whitespace.  For example, expanding this becomes:
```python
for x in range(10):_
    if x < 2:
        y = 1
    else:
        y = x - 1
```
The `_` marks the whitespace error, which is a gap left between the `for_statement` and previously inlined `body`.  The only way around this that I saw, was to leverage (1.) to replace the entire for/if statement node, but maybe the whitespace trimming will be useful to others.

3. `helpers.padded_node_text()` can now accept `padding` with values that are optionally a table instead of a string.  I did my best to document it with comments, so I won't repeat it here, since this is already a wall a text.  It receives a `prev_text` param from `toggle_multiline.lua:collapse_child_nodes()` that allows it to be semi-context aware.

The tldr reason why, is that the expected `_is_not_` and `_not_in_` would be padded as `_is__not_` and `_not__in_` and it needed to be smart enough to only have 1 space between them.  But also for unary/binary `-`, it can now pad differently for each.

### CODE REVIEW

**Comment by CKolkey** [comment]
Really great write up, and really great effort. I'll think this over, then get back to you with a plan for getting this merged :)

File: lua/ts-node-action/filetypes/python.lua [hunk 1]
```diff
@@ -1,10 +1,369 @@
...
+  local condition    = conditional_expression:named_child(cond_order[1])
+  local consequence  = conditional_expression:named_child(cond_order[2])
+  local alternative  = conditional_expression:named_child(cond_order[3])
+  local start_row, start_col = parent:start()
+  local row_parent = find_real_row_parent(parent, parent_type, start_row)
+  local cursor = {}
+  -- when we are embedded on the end of an inlined if/for statement, we need
+  -- to expand on to the next line and shift the cursor/indent
+  local if_indent   = ""
+  local else_indent = ""
```
Comment (CKolkey) [inline lua/ts-node-action/filetypes/python.lua:190]: I'm curious about this. Is `format = true` not sufficient?

**Comment by CKolkey** [comment]
Alright! Since there's a lot more complexity in here than I had to do for ruby, I figured it was a good opportunity to build the test framework. #21 implements it, and I'll be working on that this weekend when I have time. Do you also mind taking a look at the new helper I added, `destructure_node()`? I think it might be handy for teasing apart a tree into it's parts without needing to rely on the node's positions.

**Comment by CKolkey** [comment]
I merged the testing framework into master - you should be able to rebase and start using it. There are a handful of ruby examples, but since it's new just lemme know if anything about it is unclear or needs to be added :)

File: spec/filetypes/ruby_spec.lua [hunk 2]
```diff
@@ -218,4 +218,43 @@ describe("array", function()
...
+        "  2, # c",
+        "=begin",
+        "a multiline comment here",
+        "and one more line",
+        "=end",
+        "  3 # d",
+        "]"
+      })
+    )
+  end)
```
Comment (CKolkey) [inline spec/filetypes/ruby_spec.lua:237]: I'm not sure about throwing out the comment altogether. I think it would be more reasonable to just bail and not change the buffer, like when attempting to collapse a multi-line block into a single line.

File: lua/ts-node-action/init.lua [hunk 3]
```diff
@@ -19,11 +19,11 @@ local function replace_node(node, replacement, opts)
     start_row, start_col, end_row, end_col, replacement
   )
 
-  if opts.strip_whitespace then
```
Comment (CKolkey) [inline lua/ts-node-action/init.lua]: Could the same thing be accomplished by passing a callback function?

**Comment by CKolkey** [comment]
Really great work with all of this! If you feel like it's in a good spot then I'll play around with it a bit and see about getting it merged (time permitting) over the next few days :)

File: lua/ts-node-action/init.lua (hunk 3 above)
Comment (CKolkey) [reply lua/ts-node-action/init.lua]: I'll answer my own question: yes, but, I also don't think it's unreasonable to expose this as an option, as it's pretty similar to how we do formatting.

**Comment by CKolkey** [comment]
Had a chance to read this over - good lord, Python is fussy. Bravo. My only "major" note is that I'd rather not strip comments out. If we encounter a comment, I think it's more reasonable to bail than to throw the comment away.

File: lua/ts-node-action/filetypes/python.lua [hunk 4]
```diff
@@ -1,26 +1,534 @@
...
+    text:sub(1, 1) ~= "(" then
+    return "(" .. text .. ")"
+  end
+  return text
+end
+
+-- Recreating actions.toggle_multiline.collapse_child_nodes() here because
+-- it is not exported.  It was not possible to use helpers.node_text() on a
+-- multiline node because it will include the "\n", which is invalid for the
+-- replacement text.
```
Comment (CKolkey) [inline lua/ts-node-action/filetypes/python.lua:95]: Do you think it would be helpful to strip the `\n` literals from the return value of `helpers.node_text()`?  I don't see the reason to keep them. It would probably make more sense to return multi-line nodes as tables, ie, something like:
```lua
function M.node_text(node)
  local text = vim.treesitter.query.get_node_text(node, vim.api.nvim_get_current_buf())
  if text:match("\n") then
    return vim.tbl_map(vim.trim, vim.split(text, "\n"))
  else
    return text
  end
end
```

Comment (CKolkey) [reply lua/ts-node-action/filetypes/python.lua:95]: There might be some callers that are checking for the presence of `\n`, and those could be updated to just check if its a table.

**Review by CKolkey — CHANGES_REQUESTED** [review]
Some notes - altogether pretty minor changes. Really happy to see the testing story work out so well :D

**Comment by CKolkey** [comment]
Really good stuff. Let's get this merged and we can iron out any issues in a future PR. Thanks again for the contribution and effort :D
